File: tools/wasm/browser/makefile.test.mjs
import test from "node:test";
import assert from "node:assert/strict";
import fs from "node:fs";
import { parseMakefile, planMakefile } from "./makefile.mjs";

test("browser Makefiles plan Renvo commands in dependency order", () => {
  const file = parseMakefile("CFLAGS = -O2\nall: app\napp: a.o b.o\n\trenvo cc $^ -o $@\na.o: a.c\n\t@renvo cc $(CFLAGS) -c $< -o $@\nb.o: b.c\n\trenvo cc $(CFLAGS) -c $< -o $@\n");
  const plan = planMakefile(file, [], (name) => name === "a.c" || name === "b.c");
  assert.deepEqual(plan.map((item) => item.text), [
    "renvo cc -O2 -c a.c -o a.o", "renvo cc -O2 -c b.c -o b.o", "renvo cc a.o b.o -o app",
  ]);
  assert.equal(plan[0].quiet, true);
});

test("browser Makefiles reject shell commands", () => {
  assert.throws(() => planMakefile(parseMakefile("all:\n\tcc a.c\n")), /must invoke renvo/);
});

test("the Web IDE terminal runs Makefiles against the virtual workspace", () => {
  const html = fs.readFileSync(new URL("./index.html", import.meta.url), "utf8");
  const app = fs.readFileSync(new URL("./app.mjs", import.meta.url), "utf8");
  const worker = fs.readFileSync(new URL("./worker.mjs", import.meta.url), "utf8");
  assert.match(html, /id="terminal-host"/);
  assert.match(html, /@xterm\/xterm@5\.5\.0/);
  assert.match(app, /function executeTerminalLine\(line\)/);
  assert.match(app, /async function runTerminalArguments\(inputArgs\)/);
  assert.match(app, /action: "terminal"/);
  assert.match(app, /workspacePayload\(true\)/);
  assert.match(app, /syncGeneratedFiles\(result\.files\)/);
  assert.match(worker, /terminalCompilerModule/);
  assert.match(worker, /filesystemComplete: true/);
  assert.match(worker, /\["renvo", \.\.\.\(request\.args \|\| \[\]\)\]/);
  assert.match(app, /terminalRenvoArguments\(inputArgs\)/);
  assert.match(app, /command === "file"/);
  assert.match(app, /async function runTerminalWASI\(name, args = \[\]\)/);
  assert.match(app, /command\.endsWith\("\.wasm"\) \|\| command\.startsWith\("\.\/"\)/);
  assert.match(worker, /request\.type === "terminal-run"/);
  assert.match(app, /Ctrl\+A\/E\/U\/K\/W\/L/);
  assert.match(app, /attachCustomKeyEventHandler/);
  assert.match(app, /event\.key !== "Tab"[\s\S]*?event\.preventDefault\(\);[\s\S]*?event\.stopPropagation\(\);[\s\S]*?completeTerminalInput\(\);/);
  assert.match(app, /function longestCommonPrefix\(values\)/);
  assert.match(app, /if \(wasAtEnd\) xtermTerminal\.write\(data\)/);
  assert.match(app, /function replaceTerminalLine\(value\)[^]*oldCursor[^]*renderedLength/);
  assert.doesNotMatch(app, /function replaceTerminalLine\(value\)[^}]*renderTerminalInput\(\)/);
  assert.doesNotMatch(app, /if \(build\.action === "terminal"\)[^]*writeTerminal\(report\)/);
});

test("failed terminal builds resolve without replacing the validated editor build", () => {
  const app = fs.readFileSync(new URL("./app.mjs", import.meta.url), "utf8");
  assert.match(app, /if \(build\.action !== "terminal"[^]*buildValidationState/);
  assert.match(app, /terminalBuildResolve\?\.\(result\)/);
});

File: internal/makefile/makefile_test.go
package makefile

import "testing"

func TestParseAndPlanDependencyOrder(t *testing.T) {
	src := []byte("RENVO := renvo\nCFLAGS = -O2 -I include\nall: app\napp: main.o db.o\n\t$(RENVO) cc $^ -o $@\nmain.o: main.c\n\t@$(RENVO) cc $(CFLAGS) -c $< -o $@\ndb.o: db.c\n\t$(RENVO) cc $(CFLAGS) -c $< -o $@\n.PHONY: all\n")
	file, parseError := Parse(src)
	if parseError.Message != "" {
		t.Fatalf("Parse error = %#v", parseError)
	}
	existing := map[string]int64{"main.c": 1, "db.c": 1}
	commands, planError := Plan(file, nil, func(path string) (int64, bool) { stamp, ok := existing[path]; return stamp, ok })
	if planError.Message != "" {
		t.Fatalf("Plan error = %#v", planError)
	}
	if len(commands) != 3 {
		t.Fatalf("commands = %#v", commands)
	}
	want := []string{
		"renvo cc -O2 -I include -c main.c -o main.o",
		"renvo cc -O2 -I include -c db.c -o db.o",
		"renvo cc main.o db.o -o app",
	}
	for i := 0; i < len(want); i++ {
		if commands[i].Text != want[i] {
			t.Fatalf("command %d = %q, want %q", i, commands[i].Text, want[i])
		}
	}
	if !commands[0].Quiet {
		t.Fatal("@ recipe was not quiet")
	}
}

func TestRejectsShellRecipe(t *testing.T) {
	file, parseError := Parse([]byte("all:\n\tcc main.c -o app\n"))
	if parseError.Message != "" {
		t.Fatal(parseError)
	}
	_, planError := Plan(file, nil, nil)
	if planError.Message != "recipes must invoke renvo directly" {
		t.Fatalf("error = %#v", planError)
	}
}

func TestReportsDependencyCycle(t *testing.T) {
	file, _ := Parse([]byte("a: b\n\trenvo build b\nb: a\n\trenvo build a\n"))
	_, planError := Plan(file, []string{"a"}, nil)
	if planError.Message != "dependency cycle contains a" {
		t.Fatalf("error = %#v", planError)
	}
}
